### PR TITLE
Fix issue with LDAP groups in the applicable field

### PR DESCRIPTION
**Pull request by jvillafanez**
## Description
The "applicable" field in the external storage configuration wasn't working for LDAP groups. This issue happens on new installations where the configured LDAP attribute for the groupname is the "objectguid" or "entryuid".
Old installations should have the "cn" attribute configured instead, which should match the configured displayname for the group, so the issue shouldn't happen

## Related Issue
https://github.com/owncloud/core/issues/41062

## Motivation and Context
The "applicable" field should work properly regardless of the LDAP configuration. In particular, the default configuration should work.

## How Has This Been Tested?

- test environment:
- test case 1:
- test case 2:
- ...

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

File: changelog/unreleased/41063 [hunk 1]
```diff
@@ -0,0 +1,11 @@
+Bugfix: LDAP groups will be properly applied to external storages
+
+The admin can setup external storages to be used by specific users and groups.
+When a LDAP group was setup, there were some issues so users belonging to
+that group weren't be able to access to the external storage even though they
+should.
+
+Now, users belonging to LDAP groups can access to external storages configured
+to be accessed by those group.
```
Comment (phil-davis) [inline changelog/unreleased/41063:9]:
```suggestion
Now, users belonging to LDAP groups can access external storages configured
to be accessed by those groups.
```

File: changelog/unreleased/41063 [hunk 2]
```diff
@@ -0,0 +1,11 @@
+Bugfix: LDAP groups will be properly applied to external storages
+
+The admin can setup external storages to be used by specific users and groups.
+When a LDAP group was setup, there were some issues so users belonging to
+that group weren't be able to access to the external storage even though they
```
Comment (phil-davis) [inline changelog/unreleased/41063:5]:
```suggestion
that group weren't able to access the external storage even though they
```

**Review by phil-davis — COMMENTED** [review]
Just a few comments on the changelog

File: apps/files_external/lib/Controller/ApplicableController.php [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+use OCP\AppFramework\Http\JSONResponse;
+use OCP\IRequest;
+use OCP\IUserManager;
+use OCP\IGroupManager;
+
+class ApplicableController extends Controller {
+	/** @var IUserManager */
+	private $userManager;
+	/** @var IGroupManager */
+	private $groupManager;
```
Comment (DeepDiver1975) [inline apps/files_external/lib/Controller/ApplicableController.php:32]: add type declaration?

File: apps/files_external/lib/Controller/ApplicableController.php [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+		$this->groupManager = $groupManager;
+	}
+
+	/**
+	 * @param string $pattern
+	 * @param int $limit
+	 * @param int $offset
+	 * @return JSONResponse
+	 */
+	public function patternSearch($pattern = '', $limit = null, $offset = null) {
```
Comment (DeepDiver1975) [inline apps/files_external/lib/Controller/ApplicableController.php:46]: add return type

Comment (DeepDiver1975) [reply apps/files_external/lib/Controller/ApplicableController.php:46]: will type declarations work on controller methods?

File: apps/files_external/lib/Controller/ApplicableController.php [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+			'status' => 'success',
+		]);
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * @param array $users
+	 * @return JSONResponse
+	 */
+	public function userDisplayNames($users) {
```
Comment (DeepDiver1975) [inline apps/files_external/lib/Controller/ApplicableController.php:69]: return type

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]
Just some remarks on type declarations

File: apps/files_external/lib/Controller/ApplicableController.php [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+		parent::__construct($appName, $request);
+		$this->userManager = $userManager;
+		$this->groupManager = $groupManager;
+	}
+
+	/**
+	 * @param string $pattern
+	 * @param int $limit
+	 * @param int $offset
+	 * @return JSONResponse
```
Comment (DeepDiver1975) [inline apps/files_external/lib/Controller/ApplicableController.php:50]: `@return` is no longer needed as soon as proper return declarations are used - same for argument type declarations (when used)

File: apps/files_external/tests/Controller/ApplicableControllerTest.php [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+	protected function setUp(): void {
+		$this->request = $this->createMock(IRequest::class);
+		$this->userManager = $this->createMock(IUserManager::class);
+		$this->groupManager = $this->createMock(IGroupManager::class);
+
+		$this->controller = new ApplicableController('applicableController', $this->request, $this->userManager, $this->groupManager);
+	}
+
+	private function createMockUser($uid, $displayname) {
```
Comment (DeepDiver1975) [inline apps/files_external/tests/Controller/ApplicableControllerTest.php:49]: return declarations everywhere please :grin: 
as well as function argument types

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]

**Review by DeepDiver1975 — APPROVED** [review]

**Pull request merged**